core/src/Cache/Keys/MsalRefreshTokenCacheKey.cs

```diff
@@ -0,0 +1,125 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Text;
+
+namespace Microsoft.Identity.Core.Cache
+{
+    /// <summary>
+    /// An object representing the key of the token cache RT dictionary. The 
+    /// format of the key is not important for this library, as long as it is unique.
+    /// </summary>
+    internal class MsalRefreshTokenCacheKey
+    {
+        private readonly string _environment;
+        private readonly string _homeAccountId;
+        private readonly string _clientId;
+
+        internal MsalRefreshTokenCacheKey(string environment, string clientId, string userIdentifier)
+        {
+            if (string.IsNullOrEmpty(environment))
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            if (string.IsNullOrEmpty(clientId))
+            {
+                throw new ArgumentNullException(nameof(clientId));
+            }
+
+            _environment = environment;
+            _homeAccountId = userIdentifier;
+            _clientId = clientId;
+        }
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+
+            stringBuilder.Append(_homeAccountId ?? "");
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(this._environment);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(MsalCacheConstants.RefreshToken);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_clientId);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            return stringBuilder.ToString();
+        }
+
+        #region iOS
+
+        public string GetiOSAccountKey()
+        {
+            var stringBuilder = new StringBuilder();
+
+            stringBuilder.Append(_homeAccountId ?? "");
```

Comment:
why would we not have homeAccountId for MSAL tokens (new schema)?

Reply:
Existing code was like this. I see 2 possibilities for this confidential client flows that does not involve client and ADAL v3 where we do not have the idtoken.

---
In reply to: [221721468](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1287#discussion_r221721468) [](ancestors = 221721468)